### PR TITLE
SOAgradebook

### PR DESCRIPTION
**Pull request by shahn786**
added changes for gradebook team

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -58,6 +58,19 @@
...
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-amqp</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.amqp</groupId>
+			<artifactId>spring-rabbit-test</artifactId>
+			<scope>runtime</scope>
+		</dependency>
+		<!-- end rabbitmq -->
+
```
Comment (MP-CSUMB) [inline pom.xml:73]: Sets RabbitMQ dependency.

File: src/main/java/com/cst438/controller/AssignmentController.java [hunk 2]
```diff
@@ -14,27 +15,56 @@
...
+            dlist.add(new AssignmentDTO(
+                    a.getAssignmentId(),
+                    a.getTitle(),
+                    a.getDueDateAsString(),
+                    a.getSection().getCourse().getCourseId(),
+                    a.getSection().getSecId(),
+                    a.getSection().getSectionNo()));
+        }
+
+        return dlist;
```
Comment (MP-CSUMB) [inline src/main/java/com/cst438/controller/AssignmentController.java]: Just reformatting lines for clarity.

File: src/main/java/com/cst438/controller/EnrollmentController.java [hunk 3]
```diff
@@ -1,42 +1,95 @@
...
+                    e.getSection().getBuilding(),
+                    e.getSection().getRoom(),
+                    e.getSection().getTimes(),
+                    e.getSection().getCourse().getCredits(),
+                    e.getSection().getTerm().getYear(),
+                    e.getSection().getTerm().getSemester()
+            );
 
+            registrarService.updateEnrollmentGrade(enrollmentDTO);
+        }
```
Comment (MP-CSUMB) [inline src/main/java/com/cst438/controller/EnrollmentController.java]: Correctly packages the enrollment information into the enrollmentDTO to send in a message to the registrar to update the final grade of the student.

File: src/main/java/com/cst438/domain/Term.java [hunk 4]
```diff
@@ -1,14 +1,13 @@
...
-import jakarta.persistence.Entity;
-import jakarta.persistence.Id;
+import jakarta.persistence.*;
 
 import java.sql.Date;
 
 @Entity
 public class Term {
     @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (MP-CSUMB) [inline src/main/java/com/cst438/domain/Term.java]: Mistake, should be removed. Will fix.

File: src/main/java/com/cst438/service/RegistrarServiceProxy.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    CourseRepository courseRepository;
+
+    @Autowired
+    SectionRepository sectionRepository;
+
+    @Autowired
+    UserRepository userRepository;
+
+    @Autowired
+    EnrollmentRepository enrollmentRepository;
```
Comment (MP-CSUMB) [inline src/main/java/com/cst438/service/RegistrarServiceProxy.java]: Wires the repositories and Rabbit correctly.

File: src/main/java/com/cst438/service/RegistrarServiceProxy.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+    // Update Final Grades
+    public void updateEnrollmentGrade(EnrollmentDTO enrollment) {
+        sendMessage("updateEnrollment " + asJsonString(enrollment));
+    }
```
Comment (MP-CSUMB) [inline src/main/java/com/cst438/service/RegistrarServiceProxy.java:48]: Correctly sends message to update the final grade in the registrar's database.

File: src/main/java/com/cst438/service/RegistrarServiceProxy.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                enrollmentRepository.deleteById(Integer.parseInt(parts[1]));
+
+            } else {
+                System.out.println("Invalid action: " + key);
+            }
+
+        } catch (Exception e) {
+            System.out.println("Exception in receivedFromRegistrar " + e.getMessage());
+        }
+    }
```
Comment (MP-CSUMB) [inline src/main/java/com/cst438/service/RegistrarServiceProxy.java]: Correctly listens for the various action methods in the registrar's code. They correspond 1 send and 1 receive.

File: src/main/java/com/cst438/service/RegistrarServiceProxy.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+    private static <T> T  fromJsonString(String str, Class<T> valueType ) {
+        try {
+            return new ObjectMapper().readValue(str, valueType);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
```
Comment (MP-CSUMB) [inline src/main/java/com/cst438/service/RegistrarServiceProxy.java:178]: Correctly sends the messages to the registrar. This method also converts the JSON to a string in order to send it as a message (String).

File: src/main/resources/application.properties [hunk 9]
```diff
@@ -1,3 +1,4 @@
+server.port=8081
```
Comment (MP-CSUMB) [inline src/main/resources/application.properties:1]: Establishes gradebook on port 8081.

**Review by MP-CSUMB — APPROVED** [review]

**Review by chzavala — APPROVED** [review]
Everything is working no errors

**Pull request closed without merging**